### PR TITLE
Staging: Fix reading from stdin + support additional delimiters

### PR DESCRIPTION
**Pull request by rbehjati**

### CODE REVIEW

File: src/filesystem.cc [hunk 1]
```diff
@@ -90,7 +90,12 @@ class PosixReadableFile : public ReadableFile {
   
   bool ReadLine(absl::string_view *line) {
     if (mem_ == nullptr) {
-      return static_cast<bool>(std::getline(std::cin, lines_.emplace_back(), delim_));
+      std::string temp;
+      if (std::getline(std::cin, temp, delim_)) {
+        *line = absl::string_view(temp);
+        return true;
+      }
+      return false;
```
Comment (rbehjati) [inline src/filesystem.cc:98]: @vmarkovtsev @yiyangh-ps @kuba: I've made this change hoping to enable reading from stdin in spm_encode. And it seems to work, but each time I run it, the output is slightly different! 

Do you have any ideas why this is happening? 

Here are a few example runs:

*Run 1:*
```
# echo "I saw a girl with a telescope." | ./build/src/spm_encode --model=/model/pile_100k.sp.model --output_format=piece                   
filesystem.cc(96) LOG(INFO) Read: I saw a girl with a telescope.

▁@ $ <0xDB> <0xAA> <0xEF> <0xBF> <0xBD> <0x00> <0x00> girl ▁with ▁a ▁telescope .
spm_encode_main.cc(200) LOG(INFO) Encoded 1 sentences
spm_encode_main.cc(224) LOG(INFO) Encoded 1 sentences
```

*Run 2:*
```
# echo "I saw a girl with a telescope." | ./build/src/spm_encode --model=/model/pile_100k.sp.model --output_format=piece
filesystem.cc(96) LOG(INFO) Read: I saw a girl with a telescope.

▁@ <0xEF> <0xBF> <0xBD> <0xEF> <0xBF> <0xBD> <0xD3> <0xAA> <0xEF> <0xBF> <0xBD> <0x00> <0x00> girl ▁with ▁a ▁telescope .
spm_encode_main.cc(200) LOG(INFO) Encoded 1 sentences
spm_encode_main.cc(224) LOG(INFO) Encoded 1 sentences
```

*Run 3:*

```
# echo "I saw a girl with a telescope." | ./build/src/spm_encode --model=/model/pile_100k.sp.model --output_format=piece
filesystem.cc(96) LOG(INFO) Read: I saw a girl with a telescope.

▁@ <0xEF> <0xBF> <0xBD> <0xEF> <0xBF> <0xBD> <0xE7> <0x9D> <0x80> <0x00> <0x00> girl ▁with ▁a ▁telescope .
spm_encode_main.cc(200) LOG(INFO) Encoded 1 sentences
spm_encode_main.cc(224) LOG(INFO) Encoded 1 sentences
```

I guess there is some garbage in the memory somewhere that I am putting in the `line` in line 95, but I am not sure where it is coming from.

Comment (rbehjati) [reply src/filesystem.cc]: Problem solved... thanks to @joerowell :)

**Review by yiyangh-ps — CHANGES_REQUESTED** [review]

**Review by yiyangh-ps — CHANGES_REQUESTED** [review]

**Review by yiyangh-ps — APPROVED** [review]

**Comment by rbehjati** [comment]
Update: This is not really fixed yet. On larger files, in DEBUG mode, `spm_encode` fails with:
```
spm_encode: src/mixed_text_code_handler.cc:43: std::optional<sentencepiece::MixedTextCodeIterator::BlockType> sentencepiece::MixedTextCodeIterator::ReadCodeBlock(std::string_view*): Assertion `*head_ == verbatim_control_char_' failed.
Aborted (core dumped)
```

On release mode, it fails with:

```
terminate called after throwing an instance of 'std::length_error'
  what():  basic_string::_M_create
Aborted
```

I believe this is because the `mem_` buffer gets overwritten before all process threads are done processing `string_view`s from it.

File: src/mixed_text_code_handler.cc [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    in_text_ = false;
+  }
+  return line->size() > 0 ? std::make_optional(BlockType::Text) : std::nullopt;
+}
+
+std::optional<MixedTextCodeIterator::BlockType> MixedTextCodeIterator::ReadCodeBlock(absl::string_view* line) {
+  assert(*head_ == verbatim_control_char_);
+  auto ptr = reinterpret_cast<const char *>(memchr(head_, code_block_end_, tail_ - head_));
+  assert((void("Code block does not end with code block end character"), ptr != nullptr));
+  *line = absl::string_view(head_, ptr - head_);
```
Comment (kuba--) [inline src/mixed_text_code_handler.cc:46]: is it OK?
actually we subtract `head_` two times, so if `ptr` is not > `head_` it's gonna seg. fault.

File: src/filesystem.cc [hunk 3]
```diff
@@ -89,12 +99,25 @@ class PosixReadableFile : public ReadableFile {
   }
   
   bool ReadLine(absl::string_view *line) {
-    if (mem_ == nullptr) {
-      return static_cast<bool>(std::getline(std::cin, lines_.emplace_back(), delim_));
-    }
     size_t size_left = file_size_ - (head_ - mem_);
     if (size_left == 0) {
-      return false;
+      if (stdin_) {
```
Comment (vmarkovtsev) [inline src/filesystem.cc:104]: This code is not thread-safe, ofc

Comment (rbehjati) [reply src/filesystem.cc:104]: No unfortunately this part is not thread-safe. 

Honestly, I would have left the code single-threaded, and use it in a map-reduce style instead of making it multi-threaded. I guess encoding is stateless, so a map-reduce style invocation should work, and it would be very performant. The only extra overhead is loading the model multiple times.

File: src/mixed_text_code_handler.cc (hunk 2 above)
Comment (rbehjati) [reply src/mixed_text_code_handler.cc:46]: I have not looked very closely at this part of the code, its brought here from $4. But I don't think this is the source of the problem.

**Comment by rbehjati** [comment]
> In Debug mode the assert macro works, so it means:
> 
> ```
> assert(*head_ == verbatim_control_char_);
> ```
> 
> this is not true. And if it's seg fault probably `head_` is nullptr (I guess), so after dereferencing it crashes.

It is not a segmentation fault. I think the character at `head_` has changed (i.e., replaced by a new byte from stdin) by the time we reach the `assert`!

File: src/filesystem.cc [hunk 4]
```diff
@@ -89,12 +99,25 @@ class PosixReadableFile : public ReadableFile {
...
+        auto bytes_read = read(STDIN_FILENO, mem_, STDIN_BLOCK_SIZE);
```
Comment (joerowell) [inline src/filesystem.cc:105]: I think this line is what's responsible for your failing assert:

- `absl::string_view` is a view into the underlying `mem_` buffer. That is, `absl::string_view` doesn't own the memory.
- Let's suppose there's two threads (`T0, T1`) that are running. `T0` acquires a `line` that reaches the end of the buffer (that is, `size_left == 0`). `line` will be a view to a valid string inside `mem_` from the perspective of `T0`. 
- Now, suppose that `T1` now wishes to obtain a new line (and it sees that `size_left == 0`). The buffer is empty, and thus `T1` reads a new line in from `stdin`. 

At this stage, the `line` held by `T0` is no longer guaranteed to be valid, because the underlying memory has been overwritten. I think this is actually undefined behaviour, because the behaviour of what will happen depends on the underlying machine architecture. 

To put it differently, I think you're right in saying that `*head` changes. The best way to check this is to try to run the code with a single thread and to see if it works.

Comment (rbehjati) [reply src/filesystem.cc:105]: Yes. That line is exactly where the buffer content is overwritten.

I did not try it with a single thread. It requires changing `spm_encode_main.cc` to call `processChunk` directly instead of using `pool.Schedule`. 

I am not sure if it is worth experimenting with this any more. I am thinking of refactoring the main logic into a separate class, and use that in Rust via FFI. That sounds like a more viable solution to me right now :)

Comment (vmarkovtsev) [reply src/filesystem.cc:105]: I suggest wrapping the whole function in a mutex, and the problem should be solved without FFI-s and Rust-s. It takes 95% of the time to execute the model instead of to read the file. We'll be able to optimize it later.

Comment (vmarkovtsev) [reply src/filesystem.cc:105]: We are working in a startup. Time right now is much more expensive than time in the future. It is totally fine to apply a dirty hack and revisit the whole thing later when we ship the first version of the bloody dataset.

Comment (rbehjati) [reply src/filesystem.cc:105]: We already have a dirty hack. We can pass the files as input, using the `--input` flag, and use `spm_encode` the same way we plan to use `spm_train`. The only problem is that the resulting files will be much smaller than our target size, i.e., they will be about 4GB compressed, instead of 40GB compressed. We can solve this problem by merging the files together. It means more I/O, but that sounds negligible given your 95% estimate above for model execution (I guess you mean encoding the content) vs loading the files. 

IIUC, sentencepiece is something we plan to use long term. Given that we already have a dirty hack for the first version of the dataset, I suggest we start investing on the repo for our long-term use.

Comment (rbehjati) [reply src/filesystem.cc:105]: > I suggest wrapping the whole function in a mutex, and the problem should be solved without FFI-s and Rust-s.

I cannot imagine how just a mutex can solve the problem. The main thread, which is responsible for reading from stdin, could acquire the lock and overwrite the the content of `mem_` while there are still other threads holding `string_view`s based on the old content. I believe `string_view` is not the correct abstraction for communication between the main thread (which reads from stdin) and the processing threads. We need something that can communicate to the main thread when it is safe to read new bytes from stdin.

We probably need a wrapper around a `string_view` with some metadata:

```cpp
struct SyncedStringView {
  string_view line_;
  // ReadLine sets it to false; the processing thread sets it to true when it is done with `line_`.
  std::atomic<bool> done_;
}
```

Then `PosixReadableFile` could keep a vector of these objects, and whenever `size_left` is zero it has to wait for all  elements to get `done_ == true`, remove them from the vector, and start anew.

Comment (yiyangh-ps) [reply src/filesystem.cc:105]: We could use `typedef std::variant<absl::string_view, std::shared_ptr<std::string>> ps_string` to pass around instead of `absl::string_view`. Then get the pointer to the data buffer and size with:

```c++
std::tuple<const char*, size_t> GetBufferAndLength(ps_string str) {
  if (auto sv = std::get_if<absl::string_view>(&str); sv != nullptr) {
    return std::make_tuple(sv->data(), sv->length());
  }
  auto& typed_str = std::get<std::string>(str);
  return std::make_tuple(typed_str.data(), typed_str.length());
}

// Then use it like:
void some_function(ps_string line) {
  auto [buffer, len] = line;
  ...
}
```

Comment (rbehjati) [reply src/filesystem.cc:105]: Thanks @yiyangh-ps. Using `std::variant` is a great idea. I now have a fix, and have tested it on a large (16 GB) file, without problem. The performance is similar to reading from file, which should not be surprizing since the compute heavy part is the execution of the model.

File: src/filesystem.h [hunk 5]
```diff
@@ -20,13 +20,16 @@
...
+#include <variant>
 
 #include "common.h"
 #include "sentencepiece_processor.h"
 #include "absl/strings/string_view.h"
 
 namespace sentencepiece {
 namespace filesystem {
+
+typedef std::variant<absl::string_view, std::string> ps_string;
```
Comment (yiyangh-ps) [inline src/filesystem.h:32]: I would avoid directly using `std::string` because copying such variant means copying the underlying string buffer. Like I suggested earlier, `std::variant<..., std::shared_ptr<std::string>>` has a smaller copy overhead. You can create a such variant with:

```c++
ps_string* line;
std::shared_ptr<std::string> tmp = std::make_shared<std::string>(); // This allocates the buffer
std::getline(std::cin, *tmp, delim_);
*line = tmp;
```

Comment (rbehjati) [reply src/filesystem.h:32]: Where is the memory allocated in this case, and how is it freed?

Comment (yiyangh-ps) [reply src/filesystem.h:32]: [Smart pointer](https://learn.microsoft.com/en-us/cpp/cpp/smart-pointers-modern-cpp?view=msvc-170)

**Comment by vmarkovtsev** [comment]
Oh no, I see `std::getline` again. Please, please, don't use it. This is a common C++ antipattern.
- https://news.ycombinator.com/item?id=7053375
- https://gcc.gnu.org/bugzilla/show_bug.cgi?id=45574
- https://julien.jorge.st/posts/en/effortless-performance-improvements-in-cpp-std-string_view/
- My own experience since 2012

If we struggle to make my original code work, let me spend an hour solving the memory management problem. As I understand, Yiyang's proposal of `std::variant` should fix the original problem.

File: src/filesystem.h (hunk 5 above)
Comment (rbehjati) [reply src/filesystem.h:32]: That link did not have a single mention of `make_shared`! But if it allocates memory on heap, and takes care of freeing, then it would do. 
I read that `std::shared_ptr` is the equivalent of `Arc` in Rust. If that is the case. Then it can be used.

**Comment by rbehjati** [comment]
> Oh no, I see `std::getline` again. Please, please, don't use it. This is a common C++ antipattern.
> 
> * https://news.ycombinator.com/item?id=7053375
> * https://gcc.gnu.org/bugzilla/show_bug.cgi?id=45574
> * https://julien.jorge.st/posts/en/effortless-performance-improvements-in-cpp-std-string_view/
> * My own experience since 2012
> 
> If we struggle to make my original code work, let me spend an hour solving the memory management problem. As I understand, Yiyang's proposal of `std::variant` should fix the original problem.

I am not sure if that is necessary. No matter how slow `getline` is, it will be dwarfed by the model execution. In my experiments with the 16GB file, I had to slow down the main thread that calls `ReadLineStdin`, by adding a `sleep`. Without it I got OOM error, because the main thread was a lot faster in reading, than the other threads were at consuming. 

But I agree with you that in general we don't want `getline` to become the bottleneck, particularly, when we give the program more memory, and more threads.

**Comment by vmarkovtsev** [comment]
The profile is the ultimate truth ofc, I fully understand. Just that I am very suspicious that if we spawn 128 threads on a fat machine, the IO will be one of the bottlenecks.

If we see that we do well with many threads despite `std::getline` which does memory copies, bounds checks, dumb buffer reallocations, etc., etc., then we can probably ignore `std::shared_ptr`, too, and simply return `std::string`.

**Comment by rbehjati** [comment]
There is a bigger problem with using `read` vs `getline`: with `read`, whenever we fill the buffer, we probably stop reading in the middle of a document. The rest of that document will be read in the next round. This essentially means that the first part of that document will be processed and encoded by one thread, and the other part by another thread. This messes up the entire output content, as we need the entire document to be processed, and delimited as a single consecutive chunk. 

I think the fix is to change how we handle a null pointer from `memchr`. In that case, we have to move the remaining content of the buffer to its head, and only fill the rest of it.

I personally prefer, the `getline` approach with `std::variant` and `std::make_shared` (if it actually does the memory management correctly). I think it is still worth using `std:shared_prt` instead of `std::string`, because it can speedup the processing threads.

File: src/spm_encode_main.cc [hunk 6]
```diff
@@ -331,8 +336,20 @@ int main(int argc, char *argv[]) {
...
+        auto& data = std::get<std::shared_ptr<std::string>>(line);
+        counter += data->length();
+      }
       if (chunk.size() == thread_chunk_size) {
         processChunk(chunk);
+        auto diff = counter - 2147483648;
+        if (diff > 0) {
+          // Slow down a bit, if more than 2GiB worth of chunks are waiting to be processed.
+          LOG(INFO) << "Too much memory usage, waiting for " << diff / (500 * num_threads) << " ms";
+          std::this_thread::sleep_for(std::chrono::milliseconds(diff / (500 * num_threads)));
```
Comment (joerowell) [inline src/spm_encode_main.cc:351]: Two things here:

a) It's possible that we'll end up using twice as much memory as the limit here. This is because if we process a chunk and the counter is e.g `2147483647`, then this check will fail. This doesn't really matter, but it's worth keeping in mind if we end up increasing the limit to something larger. 

b) For future (and not necessary for merging this PR): we can circumvent this heuristic by using a `std::condition_variable`  check that wakes the main thread when the number of outstanding bytes drops below a certain threshold. 

That is, if we add:

```
std::mutex mutex; // The mutex for the condition variable, which isn't thread safe.
std::condition_variable cond_var; // The condition variable.
constexpr auto threshold = 2147483648; // Just to replace the magic number, but there's no need for it to be constexpr.
```

And if we replace `counter -= size` with:

```
if(counter.fetch_sub(size) < threshold + size) {
    std::unique_lock<std::mutex> lock(mutex);
    cond_var.notify_one();
}
``` 

Then we can replace the sleep with this:

```
std::unique_lock<std::mutex> lock(mutex);
cond_var.wait(lock, [&]() { return counter < 2147483648; });
```

And then the main thread will sleep until that condition is reached.

File: src/spm_encode_main.cc [hunk 7]
```diff
@@ -76,6 +76,7 @@ ABSL_FLAG(int, code_meta_block_end, -2,
   if (name == -2) name = sp.model_proto()->trainer_spec().name();
 
 int main(int argc, char *argv[]) {
+  std::atomic<unsigned> counter;
```
Comment (joerowell) [inline src/spm_encode_main.cc:79]:
```suggestion
  std::atomic<unsigned> counter{0};
```

This needs to be explicitly initialised, or the value is undefined when the program starts if you're using a pre C++20 compiler. We're currently using [C++17](https://github.com/poolsideai/sentencepiece/blob/dc661b824d74ce2aa6d21d025b75cbefa21dfedd/CMakeLists.txt#L47).

In C++17, the [standard](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/n4713.pdf) says this:

```
[See atomics.types.operations]
atomic() noexcept = default;
2 Effects: Leaves the atomic object in an uninitialized state. [ Note: These semantics ensure compatibility
with C. — end note ]
```

Whereas in C++20 and later, [it](https://isocpp.org/files/papers/N4860.pdf) says this:

```
[See atomics.types.operations]
constexpr atomic() noexcept(is_nothrow_default_constructible_v<T>);
1 Mandates: is_default_constructible_v<T> is true.
2 Effects: Initializes the atomic object with the value of T(). Initialization is not an atomic operation
(6.9.2).
```
That is, right now we're relying on undefined behaviour. Most reasonable compilers will default to 0, but that's only the case _right now_.

File: src/filesystem.h [hunk 8]
```diff
@@ -29,7 +29,7 @@
 namespace sentencepiece {
 namespace filesystem {
 
-typedef std::variant<absl::string_view, std::string> ps_string;
+typedef std::variant<absl::string_view, std::shared_ptr<std::string>> ps_string;
```
Comment (joerowell) [inline src/filesystem.h:32]:
```suggestion
using ps_string = std::variant<absl::string_view, std::shared_ptr<std::string>>;
```

`typedef` is entirely superseded by `using` (i.e everything `typedef` can do, `using` does, and more).

File: src/filesystem.cc [hunk 9]
```diff
@@ -132,8 +132,8 @@ class PosixReadableFile : public ReadableFile {
 
   bool ReadLineStdin(ps_string *line) {
     if (stdin_) {
-      std::string tmp;
-      auto worked = static_cast<bool>(std::getline(std::cin, tmp, delim_));
+      std::shared_ptr<std::string> tmp = std::make_shared<std::string>();
+      auto worked = static_cast<bool>(std::getline(std::cin, *tmp, delim_));
```
Comment (joerowell) [inline src/filesystem.cc:136]:
```suggestion
const bool worked = std::getline(std::cin, *tmp, delim_);
```

`std::getline` has a conversion operator to `bool`, which means we can avoid the cast. The semantics are the same.

**Review by joerowell — COMMENTED** [review]
I've added a few small tweaks. Only the atomic initialisation and type is important, and the rest are mostly style (feel free to ignore those).

File: src/spm_encode_main.cc (hunk 6 above)
Comment (rbehjati) [reply src/spm_encode_main.cc:351]: > a) .... This doesn't really matter, but it's worth keeping in mind if we end up increasing the limit to something larger.

Yes, we have to be careful with increasing the limit (`2147483648`), also because the `chunk` is not the only allocated memory. In my experiments, memory usage was around 4GiB when the first calls to `sleep` happened.

Using `std::condition_variable` is a good idea. We should get back to it in a future PR. Thanks :)

**Review by rbehjati — COMMENTED** [review]
Thanks for the review, and your suggestions.

**Comment by joerowell** [comment]
> I get a "no suitable conversion" error.

Huh, weird. Feel free to ignore it, then :)
